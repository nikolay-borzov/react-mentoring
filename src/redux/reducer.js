import { combineReducers } from 'redux'
import search from './modules/search'
import films from './modules/films'

const rootReducer = combineReducers({
  search,
  films
})

export default rootReducer
