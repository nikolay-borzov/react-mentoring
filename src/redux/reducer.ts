import { combineReducers } from 'redux'
import { all } from 'redux-saga/effects'

import search, { searchSagas, SearchState } from './modules/search'
import view, { viewSagas, ViewState } from './modules/view'

export interface AppState {
  search: SearchState
  view: ViewState
}

const rootReducer = combineReducers({
  search,
  view
})

export function* rootSaga() {
  yield all([searchSagas(), viewSagas()])
}

export default rootReducer
