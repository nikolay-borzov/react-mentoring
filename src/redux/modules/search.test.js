import reducer, { setParams, actionTypes, selectors } from './search'

describe('selectors', () => {
  const state = {
    search: {
      search: '',
      searchBy: 'Gemini',
      sortBy: 'title',
      sortOrder: 'desc',
      limit: 15
    }
  }

  it(`returns 'search' param`, () => {
    expect(selectors.search(state)).toBe(state.search.search)
  })

  it(`returns 'searchBy' param`, () => {
    expect(selectors.searchBy(state)).toBe(state.search.searchBy)
  })

  it(`returns 'sortBy' param`, () => {
    expect(selectors.sortBy(state)).toBe(state.search.sortBy)
  })

  it(`returns search params`, () => {
    expect(selectors.params(state)).toBe(state.search)
  })
})

describe('actions', () => {
  it('creates an action to set search params', () => {
    const payload = {
      search: 'Gemini',
      searchBy: 'title'
    }
    const expectedAction = {
      type: actionTypes.SET_PARAMS,
      payload
    }

    expect(setParams(payload)).toEqual(expectedAction)
  })

  it(`accepts only 'search', 'searchBy' and 'sortBy' params`, () => {
    const payload = {
      search: 'Gemini',
      searchBy: 'title',
      sortOrder: 'asc'
    }
    const expectedAction = {
      type: actionTypes.SET_PARAMS,
      payload: {
        search: 'Gemini',
        searchBy: 'title'
      }
    }

    expect(setParams(payload)).toEqual(expectedAction)
  })
})

describe('reducer', () => {
  it('returns initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot()
  })

  it('sets search params', () => {
    const state = { search: '', searchBy: 'genre', limit: 20 }
    const action = setParams({
      search: 'Horror',
      searchBy: 'genre'
    })

    expect(reducer(state, action)).toMatchSnapshot()
  })
})
