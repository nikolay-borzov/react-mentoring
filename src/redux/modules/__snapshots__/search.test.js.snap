// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer returns initial state 1`] = `
Object {
  "limit": 15,
  "search": "",
  "searchBy": "title",
  "sortBy": "release_date",
  "sortOrder": "desc",
}
`;

exports[`reducer sets search params 1`] = `
Object {
  "limit": 20,
  "search": "Horror",
  "searchBy": "genre",
}
`;
